includes/class-wc-gateway-ppec-cart-handler.php

```diff
@@ -38,6 +43,83 @@ public function before_cart_totals() {
 	}
 
 	/**
+	 * Generates the cart for express checkout on a product level.
+	 *
+	 * @since 1.4.0
+	 */
+	public function generate_cart() {
+		if ( ! wp_verify_nonce( $_POST['nonce'], '_wc_ppec_generate_cart_nonce' ) ) {
+			wp_die( __( 'Cheatin&#8217; huh?', 'woocommerce-gateway-paypal-express-checkout' ) );
+		}
+
+		if ( ! defined( 'WOOCOMMERCE_CART' ) ) {
+			define( 'WOOCOMMERCE_CART', true );
+		}
+
+		WC()->shipping->reset_shipping();
+
+		global $post;
+
+		$product = wc_get_product( $post->ID );
+		$qty     = ! isset( $_POST['qty'] ) ? 1 : absint( $_POST['qty'] );
+
+		/**
+		 * If this page is single product page, we need to simulate
+		 * adding the product to the cart taken account if it is a
+		 * simple or variable product.
+		 */
+		if ( is_single() ) {
+			// First empty the cart to prevent wrong calculation.
+			WC()->cart->empty_cart();
+
+			if ( 'variable' === $product->get_type() && isset( $_POST['attributes'] ) ) {
+				$attributes = array_map( 'wc_clean', $_POST['attributes'] );
+
+				if ( version_compare( WC_VERSION, '3.0', '<' ) ) {
+					$variation_id = $product->get_matching_variation( $attributes );
+				} else {
+					$data_store = WC_Data_Store::load( 'product' );
+					$variation_id = $data_store->find_matching_product_variation( $product, $attributes );
+				}
+
+				WC()->cart->add_to_cart( $product->get_id(), $qty, $variation_id, $attributes );
+			} elseif ( 'simple' === $product->get_type() ) {
+				WC()->cart->add_to_cart( $product->get_id(), $qty );
+			}
+		}
+
+		WC()->cart->calculate_totals();
+
+		wp_send_json( new stdClass() );
+	}
+
+	/**
+	 * Display paypal button on the product page.
+	 *
+	 * @since 1.4.0
+	 */
+	public function display_paypal_button_product() {
+		$gateways = WC()->payment_gateways->get_available_payment_gateways();
+
+		if ( ! is_single() || ! isset( $gateways['ppec_paypal'] ) ) {
```

Comment:
not a biggie and we can skip this for. This won't work on a page with `[product_page]` shortcode.